src/vsr/replica.zig

```diff
@@ -1833,10 +1865,7 @@ pub fn Replica(
             }
 
             // TODO We can optimize this to commit into the client table reply if it exists.
```

Comment:
We need to do this now, using the existing reply if it exists in the client table.

Otherwise we would temporarily overflow our static allocation count for the client table, by having a message reference in there already and then acquiring a new one now.

Alternatively, if reusing the existing reply message is too complicated (perhaps because it might be reffed by the send queue and so in the progress of being copied to the kernel), then we should rather +1 to our total client table count, provided we know that we will immediately unref the old reply being replaced.

Reply:
> Alternatively, if reusing the existing reply message is too complicated (perhaps because it might be reffed by the send queue and so in the progress of being copied to the kernel), then we should rather +1 to our total client table count, provided we know that we will immediately unref the old reply being replaced.

We could test that `references == 1` before modifying, and when `references != 1` just replace it like usual. But I'm leaning towards just increasing the message count +1 anyway, to avoid that (rather complicated) case.

Reply:
> But I'm leaning towards just increasing the message count +1 anyway, to avoid that (rather complicated) case.

Agreed.

> The existing code can still be improved though. Since get_message is infallible, it can be moved into commit_op rather than making the caller get the message and pass it in.

That will be great, reducing the dimensionality of the `get_message` return type is simplifying the code nicely already. We had several bugs found by the VOPR around messages not being available with respect to committing or with respect to retrying a send after a timeout. Working out the static allocation like we are doing now requires care and thought, but the alternative meant more complexity and bugs at the same time. Furthermore, our understanding around resource usage is growing because the assertions are pushing us to explore the boundary between positive/negative space.

Definitely looking forward to getting a pure crash in the VOPR instead of a stuck bug if we have any further static allocation issues in future. 😄